File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Data.Char

responseFor :: String -> String
responseFor xs
    | isSilence  = "Fine. Be that way!"
    | isQuestion && isYell = "Calm down, I know what I'm doing!"
    | isQuestion = "Sure."
    | isYell     = "Whoa, chill out!"
    | otherwise  = "Whatever."
    where
        isQuestion = last(clearSpaces xs) == '?'
        isYell     = isAllCaps (clearSpaces xs)
        isSilence  = clearSpaces xs == ""

clearSpaces :: String -> String
clearSpaces str = filter (not . isSpace) str

isAllCaps :: String -> Bool
isAllCaps str = do
    let trueMap = map (isUpper) $ filter (isLetter) str
    if trueMap == []
        then False
        else and $ trueMap
    

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    input = String.trim input
    cond do
      silence?(input)  -> "Fine. Be that way!"
      question?(input) -> "Sure."
      normal?(input)   -> "Whatever."
      yell?(input)     -> "Whoa, chill out!"
      true             -> "Whatever."
    end
  end

  defp silence?(input), do: input == ""

  defp question?(input), do: String.ends_with? input, "?"

  defp yell?(input), do: input == String.upcase(input)
  
  defp normal?(input), do: input == String.downcase(input)
end